packages/eslint-plugin/src/rules/default-param-last.ts

```diff
@@ -0,0 +1,78 @@
+import { createRule } from '../util';
+import {
+  TSESTree,
+  AST_NODE_TYPES,
+} from '@typescript-eslint/experimental-utils';
+
+export default createRule({
+  name: 'default-param-last',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Enforce default parameters to be last',
+      category: 'Best Practices',
+      recommended: false,
+    },
+    schema: [],
+    messages: {
+      shouldBeLast: 'Default parameters should be last.',
+    },
+  },
+  defaultOptions: [],
+  create(context) {
+    /**
+     * checks if node is optional parameter
+     * @param {ASTNode} node node to evaluate
+     * @returns {boolean}
+     * @private
+     */
+    function isOptionalParam(node: TSESTree.Node): boolean {
+      return node.type === AST_NODE_TYPES.Identifier && node.optional === true;
+    }
+
+    /**
+     * checks if node is plain parameter
+     * @param {ASTNode} node node to evaluate
+     * @returns {boolean}
+     * @private
+     */
+    function isPlainParam(node: TSESTree.Node): boolean {
+      return !(
+        node.type === AST_NODE_TYPES.AssignmentPattern ||
+        node.type === AST_NODE_TYPES.RestElement ||
+        isOptionalParam(node)
+      );
+    }
+
+    function checkDefaultParamLast(
+      node:
+        | TSESTree.ArrowFunctionExpression
+        | TSESTree.FunctionDeclaration
+        | TSESTree.FunctionExpression,
+    ): void {
+      let hasSeenPlainParam = false;
+      for (let i = node.params.length - 1; i >= 0; i--) {
+        const param = node.params[i];
+
+        if (isPlainParam(param)) {
+          hasSeenPlainParam = true;
+          continue;
+        }
+
+        if (
+          hasSeenPlainParam &&
+          (isOptionalParam(param) ||
+            param.type === AST_NODE_TYPES.AssignmentPattern)
+        ) {
+          context.report({ node: param, messageId: 'shouldBeLast' });
+        }
+      }
+    }
+
+    return {
+      ArrowFunctionExpression: checkDefaultParamLast,
+      FunctionDeclaration: checkDefaultParamLast,
+      FunctionExpression: checkDefaultParamLast,
```

Comment:
what do you think about adding support for other function like nodes?

- TSEmptyBodyFunctionExpression
- TSDeclareFunction
- TSConstructorType
- TSConstructSignatureDeclaration
- TSFunctionType
- TSCallSignatureDeclaration

Reply:
There's no need beacuse you can't have default params in function type signatures.
Error 2371 "A parameter initializer is only allowed in a function or constructor implementation."